lib/perl/Genome/Model/Command/Admin/RemoveDiskAllocationsFromTestdb.pm

```diff
@@ -0,0 +1,308 @@
+
+    my $cache_key = "_cached_${db_or_tmpl}_handle";
+    unless ($self->{$cache_key}) {
+        my $db_name = $db_or_tmpl eq 'database'
+                        ? $self->database_name
+                        : $self->template_name;
+        my $db_host = $self->database_server;
+        my $db_port = $self->database_port;
+
+        $self->{$cache_key} = DBI->connect("dbi:Pg:dbname=${db_name};host=${db_host};port=${db_port}",
+                                           Genome::Config::get('ds_gmschema_login'),
+                                           Genome::Config::get('ds_gmschema_auth'),
+                                           { AutoCommit => 0, RaiseError => 1, PrintError => 0 });
+    }
+    return $self->{$cache_key};
+}
+
+
+sub report_allocations_to_delete {
+    my($self, @allocations) = @_;
+
+    my $kb_sum = 0;
+    foreach my $alloc ( @allocations ) {
+        $kb_sum += $alloc->kilobytes_requested;
+    }
+
+    $self->status_message('Removing %.3f GB in %d allocations.',
+                          $kb_sum / KB_IN_ONE_GB,
+                          scalar(@allocations));
+}
+
+sub delete_allocations {
+    my($self, @stripped_allocations) = @_;
+
+    my %allocation_ids = map { $_ => 1 }
+                         map { $_->id } @stripped_allocations;
+
+    # remove allocations tied to Genome::Process objects.  We'll delete those
+    # separately
+    my @processes = Genome::Process->get(disk_allocation_id => [ keys %allocation_ids ]);
+    my @allocation_ids_from_processes = map { $_->disk_allocation_id } @processes;
+    delete @allocation_ids{ @allocation_ids_from_processes };
+
+    my @allocations = Genome::Disk::Allocation->get(id => [ keys %allocation_ids ]);
+    foreach my $real_allocation ( @allocations ) {
+        $self->_print_debug_message_about_deleting_allocation($real_allocation);
+        $real_allocation->delete();
+    }
+
+    $self->_delete_processes(\@processes);
+    return 1;
+}
+
+sub _print_debug_message_about_deleting_allocation {
+    my($self, $allocation, $process) = @_;
+
+    my($scale, $divisor) = $allocation->kilobytes_requested < KB_IN_ONE_GB
+                            ? ('MB', KB_IN_ONE_MB)
+                            : ('GB', KB_IN_ONE_GB);
+
+    my $process_message = $process
+                            ? ' from Process ' . $process->id
+                            : '';
+
+    $self->debug_message('Deleting %.3f %s for allocation %s%s',
+                         $allocation->kilobytes_requested / $divisor,
+                         $scale,
+                         $allocation->id,
+                         $process_message);
+}
+
+# Genome::Process objects delete their own allocations
+sub _delete_processes {
+    my($self, $processes) = @_;
+
+    foreach my $process ( @$processes ) {
+        my $allocation = $process->disk_allocation;
+        $self->_print_debug_message_about_deleting_allocation($allocation, $process);
+        $process->delete();
```

Comment:
It's not strictly necessary, but doing it this way avoids an ugly message about a constraint violation if the Process's allocation gets deleted without deleting the Process.